packages/sandbox/src/cloudwatch_logs_monitor.ts

```diff
@@ -0,0 +1,297 @@
+import {
+  LogLevel,
+  Printer,
+  colors,
+  format,
+  printer,
+} from '@aws-amplify/cli-core';
+import { AmplifyFault } from '@aws-amplify/platform-core';
+import {
+  CloudWatchLogsClient,
+  FilterLogEventsCommand,
+  ResourceNotFoundException,
+} from '@aws-sdk/client-cloudwatch-logs';
+import fs from 'fs';
+import path from 'path';
+
+/**
+ * After reading events from all CloudWatch log groups how long should we wait to read more events.
+ *
+ * If there is some error with reading events (i.e. Throttle) then this is also how long we wait until we try again
+ */
+const SLEEP = 2_000;
+
+/**
+ * Represents a CloudWatch Log Event that will be printed to the terminal
+ */
+type CloudWatchLogEvent = {
+  /**
+   * The log event message
+   */
+  readonly message: string;
+
+  /**
+   * The name of the log group
+   */
+  readonly logGroupName: string;
+
+  /**
+   * The time at which the event occurred
+   */
+  readonly timestamp: Date;
+};
+
+/**
+ * Represents how an event from a CloudWatch Log Group will be displayed
+ * Indexed off of logGroupName for easy retrieval during logs even polling
+ */
+type CloudWatchEventDisplay = {
+  [logGroupName: string]: {
+    friendlyName: string;
+    color: (typeof colors)[keyof typeof colors];
+  };
+};
+
+type CloudWatchLogGroupToMonitor = {
+  /**
+   * The name of the log group
+   */
+  readonly logGroupName: string;
+
+  /**
+   * How to display events for this log group (friendlyName and color)
+   */
+  readonly cloudWatchEventDisplay: CloudWatchEventDisplay;
+
+  /**
+   * Start time
+   */
+  startTime: number;
+};
+
+/**
+ * Monitors CloudWatch logs and stream it to stdout or a user provided file location
+ * Consumers can activate and deactivate the monitor. The monitor on reactivation, starts
+ * streaming from the last time it was deactivated to avoid missing any logs while the monitor
+ * was deactivated.
+ */
+export class CloudWatchLogEventMonitor {
+  /**
+   * Determines from what time the logs should be streamed
+   */
+  private startTime: number;
+
+  /**
+   * Collection of all LogGroups that need to be streamed
+   */
+  private readonly allLogGroups: CloudWatchLogGroupToMonitor[] = [];
+
+  private active = false;
+
+  private printer = printer; // default stdout
+
+  /**
+   * Initializes the start time to be `now`
+   */
+  constructor(readonly cloudWatchLogsClient: CloudWatchLogsClient) {
+    this.startTime = Date.now();
```

Comment:
This is why it mostly doesn't matter for now. Because instantiation and activation are happening within a couple ms of each other. But just from an API perspective, it's a bit odd for instantiation to start a clock that will impact the behavior of the methods. Typically ctors are for injecting dependencies

Reply:
This should go to `activate` . (and if we do this it could be removed from `deactivate`).

Reply:
The use case for this is to keep track of when we "deactivated" the log streamer such that when we resume, we resume from that point onward and not when the caller asked to resume.

This helps in pausing the streaming during the deployment. As soon as the deployment is detected, we pause the log streaming, resets the start_time to now so that when we resume, we don't miss the logs during the deployment time.

Doing this way prevents the caller to maintain timings and let this module own it.

Reply:
Should we have `start`,`stop` and `pause` then ? or perhaps `start` and `pause` for now.
Perhaps this naming would be less ambiguous. I'm not sure if `deactivate` means `stop` or `pause`.

Reply:
Yup, changing to `pause`

Reply:
I was thinking about this more and I'm wondering if over time this could evolve to something like this:
```
monitor.start() // always sets startTime as now
// observe logs then pause
monitor.pause() // pauses streaming
// observe another batch of logs
monitor.resume() // un-pauses streaming from last checkpoint
// pause again and go for lunch
monitor.pause() // pause again
// after lunch resume but from current time as going for lunch would accumulate too much logs.
monitor.restart() OR another monitor.start() // to reset checkpoint to now
```

Reply:
Perhaps, but I think that's more complicated than what we have today, `start` and `pause`

Reply:
Yes. The goal of my comment was to see if we should set ourselves towards that direction. I.e. to MVP would be to have `start`, `pause` and `resume` . Or perhaps just `pause` and `resume` if we treat ctor call as implicit start.

Reply:
The `ctor` can't do a start since we don't have the `logGroups` information. Also I think `activate` captures both start/resume.

Moreover, none of this is part of public API interface, so will be easy to change whenever we need.